File: proxy/Proxy.ts
interface Image {
    display(): void;
}

class RealImage implements Image {
    private fileName: string;
    constructor(fileName: string) {
        this.fileName = fileName;
        this.loadFromDisk(fileName);
    }
    display(): void {
        console.log('Displaying ' + this.fileName);
    }
    loadFromDisk(fileName: string): void {
        console.log('Loading ' + fileName);
    }
}

class ProxyImage implements Image {
    private realImage: RealImage;
    private fileName: string;
    constructor(fileName: string) {
        this.fileName = fileName;
    }
    display(): void {
        if (this.realImage === undefined) {
            this.realImage = new RealImage(this.fileName);
        }
        this.realImage.display();
    }
}

const image: Image = new ProxyImage('test_10mb.jpg');

// image will be loaded from disk
image.display();
console.log('');
// image will not be loaded from disk
image.display();

File: proxy/Proxy.java
public class Proxy {
    private interface Image {
        void display();
    }

    private static class RealImage implements Image {
        private String fileName;

        public RealImage(String fileName) {
            this.fileName = fileName;
            loadFromDisk(fileName);
        }

        @Override
        public void display() {
            System.out.println("Displaying " + fileName);
        }

        private void loadFromDisk(String fileName) {
            System.out.println("Loading " + fileName);
        }
    }

    private static class ProxyImage implements Image {
        private RealImage realImage;
        private String fileName;

        public ProxyImage(String fileName) {
            this.fileName = fileName;
        }

        @Override
        public void display() {
            if (realImage == null) {
                realImage = new RealImage(fileName);
            }
            realImage.display();
        }
    }

    public void testImageProxy() {
        Image image = new ProxyImage("test_10mb.jpg");
        image.display();
        System.out.println("");
        image.display();
    }

    public static void main(String[] args) {
        Proxy proxy = new Proxy();
        proxy.testImageProxy();
    }
}